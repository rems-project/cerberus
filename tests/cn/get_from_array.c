

/* Simple demo of a kind of situation where ownership is obtained via an array
   that encloses an object passed by pointer. This kind of step is required by
   the buddy-allocator proof. */

int *global_array;

enum {
  global_array_width = 42,
};

/*@ function (i32) global_array_width () @*/

static inline int get_global_array_width_for_cn (void)
/*@ cn_function global_array_width; @*/
{
  return global_array_width;
}

/*@
predicate (map<i32, i32>) Global_Array (pointer p)
{
  take Arr = each (i32 i; 0i32 <= i && i < global_array_width ())
    { Owned(array_shift<int>(p, i)) };
  return Arr;
}
@*/

void set_a_pointer(int *p, int x)
/*@ accesses global_array; @*/
/*@ requires (alloc_id) global_array == (alloc_id) p; @*/
/*@ requires take Arr = Global_Array(global_array); @*/
/*@ requires let offs = ((u64)p - (u64)global_array); @*/
/*@ requires mod(offs, (u64) (sizeof<int>)) == 0u64; @*/
/*@ requires let idx = (offs / ((u64) (sizeof<int>))); @*/
/*@ requires 0u64 <= idx && idx < ((u64) (global_array_width ())); @*/
/*@ ensures take Arr2 = Global_Array(global_array); @*/
{
  /*@ extract Owned<int>, ((i32) idx); @*/
  /*@ instantiate good<int>, ((i32) idx); @*/
  *p = x;
}

int main(void)
/*@ trusted; @*/
{
  int p[1];
  set_a_pointer(p, 5);
}
