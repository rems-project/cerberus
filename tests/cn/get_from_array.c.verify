return code: 0
tests/cn/get_from_array.c:16:5: warning: experimental keyword 'cn_function' (use of experimental features is discouraged)
/*@ cn_function global_array_width; @*/
    ^~~~~~~~~~~ 
tests/cn/get_from_array.c:24:8: warning: 'each' prefers a 'u64', but 'i' has type 'i32'.
  take Arr = each (i32 i; 0i32 <= i && i < global_array_width ())
       ^
tests/cn/get_from_array.c:40:23: warning: 'focus' prefers a 'u64', but '(i32)idx' has type 'i32'.
  /*@ focus RW<int>, ((i32) idx); @*/
                      ^~~~~~~~~ 
[1/2]: get_global_array_width_for_cn -- pass
tests/cn/get_from_array.c:41:7: warning: nothing instantiated
  /*@ instantiate good<int>, ((i32) idx); @*/
      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
[2/2]: set_a_pointer -- pass
