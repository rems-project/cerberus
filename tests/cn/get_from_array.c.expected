return code: 0
tests/cn/get_from_array.c:16:5: warning: experimental keyword 'cn_function' (use of experimental features is discouraged)
/*@ cn_function global_array_width; @*/
    ^~~~~~~~~~~ 
tests/cn/get_from_array.c:24:8: warning: 'each' expects a 'u64', but 'i' with type 'i32' was provided. This will become an error in the future.
  take Arr = each (i32 i; 0i32 <= i && i < global_array_width ())
       ^
tests/cn/get_from_array.c:40:28: warning: 'extract' expects a 'u64', but '(i32)idx' with type 'i32' was provided. This will become an error in the future.
  /*@ extract Owned<int>, ((i32) idx); @*/
                           ^~~~~~~~~ 
[1/2]: get_global_array_width_for_cn -- pass
tests/cn/get_from_array.c:41:7: warning: nothing instantiated
  /*@ instantiate good<int>, ((i32) idx); @*/
      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
[2/2]: set_a_pointer -- pass
